### PR TITLE
Add syntax highlighting

### PR DESCRIPTION
**Pull request by ekosz**
This PR relies on webpack to be able to pull in CSS from a module like this. Let me know if you're okay with this method.

screen shot 2016-11-29 at 1 49 08 pm

### CODE REVIEW

**Comment by Timer** [comment]
I really like this, but I think we can do better in terms of a theme. Let's try to get Dan's input on it.
/cc @gaearon

**Comment by Timer** [comment]
As for the css require, I think that's totally acceptable since it's behavior that's allowed in CRA.

**Comment by ekosz** [comment]
Great. [There are a lot of themes to choose from](https://highlightjs.org/static/demo/), I'm sure we can find something we like. I focused on a dark color theme as I felt it would look better with the red until we change it.

**Comment by Timer** [comment]
The light syntax theme almost looks like it goes well with that red in the demo. Maybe that's a overall theme we should adopt.

My largest peeve with it is that the code box extends all the way to the edge of the screen on the right; is there some way we can make those boxes fit the code but remain uniform (flexbox?)?

**Comment by gaearon** [comment]
Can we make it match the output we show on syntax errors? We can later change both colors but I want them to be consistent. The colors are defined [here](https://github.com/facebookincubator/create-react-app/blob/5a273b2a934df7390d3c7af5f46052d49c309b1f/packages/react-dev-utils/webpackHotDevClient.js#L28-L40). The underlying ANSI codes are coming from Babel so you can find them (and the colorizer Babel uses) somewhere in Babel sources. @hzoo may be able to give a pointer.

**Comment by hzoo** [comment]
[babel-code-frame's](https://github.com/babel/babel/tree/master/packages/babel-code-frame) colors are [here](https://github.com/babel/babel/blob/e86f62b304d280d0bab52c38d61842b853848ba6/packages/babel-code-frame/src/index.js#L9-L22)

`const result = require('babel-code-frame')(rawLines, lineNumber, colNumber, { /* opts */ });`

**Comment by ekosz** [comment]
@gaearon @hzoo Thank you for that! I've updated the PR to use `babel-code-frame`, but for some reason I can't get it to highlight. As far as I can tell it *should* work, but for some reason `babel-code-frame` isn't inserting the ANSI markup I thought it would.

When one of you get a chance could you quickly look over what I'm doing and point out anything I'm missing? Thanks!

File: packages/react-dev-utils/failFast.js [hunk 1]
```diff
@@ -164,15 +168,13 @@
       elem.appendChild(elemLink)
 
       if (!internalUrl && sourceLines.length !== 0) {
+        const sourceCode = sourceLines.map(({ text }) => text).join('\n')
+        const ansiHighlight = codeFrame(sourceCode, 6, 0, { highlightCode: true });
```
Comment (ekosz) [inline packages/react-dev-utils/failFast.js]: Hard coded line number for now

**Comment by ekosz** [comment]
@hzoo Nevermind I think I figured it out. I think `babel-code-frame` expects to be given the entire JS file so that it can properly tokenize/parse it while we only have a few lines of code to work with. I assume because if that it doesn't know where to add the highlighting. I can switch back to `highlight.js` and just try to create a stylesheet that matches `babel-code-frame` rules.

**Comment by Timer** [comment]
@ekosz we have the full source available. we can expose that in stackframe-resolve.

**Comment by Timer** [comment]
Better yet maybe add some consumer model to that package... so we can give it a method that'll return babel code frames

**Comment by ekosz** [comment]
@Timer Ah, you're right. I've bumped up the `CONTEXT` to `Infinity` and it's still not highlighting. I'll go some code spelunking some more to see whats happening.

**Comment by ekosz** [comment]
@Timer Looks like it's the `chalk` library getting in the way. Because we're calling it from the browser it never adds ANSI as it doesn't have access to Node variables. Looks like we could set `process.env.FORCE_COLOR` to make it work.

**Comment by ekosz** [comment]
@Timer @gaearon Got it working with the `babel-code-frame`, now all of the styles/colors are shared!

screen shot 2016-11-29 at 9 22 45 am

**Comment by gaearon** [comment]
Great! Can you try porting both windows to use https://github.com/facebookincubator/create-react-app/pull/1101#issuecomment-263641886 just for the sake of trying?

**Comment by ekosz** [comment]
@gaearon Yeah that's something I can try out. Will have to wait until later today once I'm out of the office.

**Comment by ekosz** [comment]
@gaearon @Timer @nicinabox Added Base16 Github styling. I did the best I could trying to match up all the rules being used.

screen shot 2016-11-29 at 12 37 42 pm

**Comment by nicinabox** [comment]
Huh, that did not colorize as much as I expected.

**Comment by ekosz** [comment]
@nicinabox `babel-code-frame` doesn't highlight many things. Just these: https://github.com/babel/babel/blob/e86f62b304d280d0bab52c38d61842b853848ba6/packages/babel-code-frame/src/index.js#L9-L22

**Comment by ekosz** [comment]
I had to update the look of the syntax error screen as well due to the change of a dark color scheme becoming light as well:

screen shot 2016-11-29 at 12 49 09 pm

**Comment by ekosz** [comment]
But if we're happy with these colors, this is ready to roll.

**Comment by gaearon** [comment]
We would need some help from @nicinabox to figure out how to make them look consistent. I'd prefer if they were more similar (except that syntax overlay shouldn't be closeable).

**Comment by ekosz** [comment]
@nicinabox @Timer @gaearon Rebased with new style.

screen shot 2016-11-29 at 1 49 08 pm

**Comment by nicinabox** [comment]
Let me think about that compiler error design. I'm about to wrap up here at the office for the day and I can focus on it when I get home.

Re rebased style: It'd be slick to pull have the white extend to the edges. I had it working with `:after` but there was no css file at the time.

**Comment by hzoo** [comment]
@nicinabox we previously choose to use less colors in the update to babel-code-frame https://github.com/babel/babel/pull/4572 - also the background was assumed to be dark like in the screenshots

**Comment by nicinabox** [comment]
TBH I'm not in love with the syntax highlighting here. Maybe it's just the colors.

**Comment by Timer** [comment]
@nicinabox I think the highlighting is important to have ... let me see what we can do about making it look better.

**Comment by ekosz** [comment]
@Timer @nicinabox What's nice about this approach is all of the color info is in one file and will update all colors to be consistent. This can probably be merged in its current form then we can mess with the colors on Timer/redbox

**Comment by Timer** [comment]
I think the white background is what turned me off ... how about this?
What do you think @nicinabox @ekosz?

**Comment by Timer** [comment]
I'm going to bring it in for now, we can continue discussion over at https://github.com/facebookincubator/create-react-app/pull/1101 and receive input from more individuals than just us.

Thanks for all your work and input, @ekosz @nicinabox!

**Comment by nicinabox** [comment]
Ahh yeah that's better.

**Pull request merged**